python/ray/dag/compiled_dag_node.py

```diff
@@ -386,6 +379,116 @@ def __init__(
             assert not isinstance(val, ChannelInterface)
             assert not isinstance(val, DAGInputAdapter)
 
+        # Input reader to read input data from upstream DAG nodes.
+        self.input_reader: ReaderInterface = SynchronousReader(self.input_channels)
+        # Output writer to write output data to downstream DAG nodes.
+        self.output_writer: WriterInterface = SynchronousWriter(self.output_channel)
+        # Store the intermediate result of a READ or COMPUTE operation.
+        # The result of a READ operation will be used by a COMPUTE operation,
+        # and the result of a COMPUTE operation will be used by a WRITE operation.
+        self._cache = None
+
+    def cancel(self):
+        self.input_reader.close()
+        self.output_writer.close()
+
+    def prepare(self):
+        """
+        Prepare the task for execution. The `exec_operation` function can only
+        be called after `prepare` has been called.
+        """
+        for typ_hint in self.input_type_hints:
+            typ_hint.register_custom_serializer()
+        self.output_type_hint.register_custom_serializer()
+        self.input_reader.start()
+        self.output_writer.start()
+
+    def set_cache(self, data):
```

Comment:
should we always require to reset cache before calling `set_cache`? otherwise, we can have unexpected overwrite? 

```
assert self._cache is None
...
```

Reply:
Updated https://github.com/ray-project/ray/pull/46911/commits/3cb1eb0ada4c27dfaa703203ab1b95cf4660bcd2